### PR TITLE
Compile fscache modules to speedup NFS shares

### PR DESCRIPTION
**Pull request by paolomainardi**
This is needed in order to speedup NFS connections, as described here: http://chase-seibert.github.io/blog/2014/03/09/vagrant-cachefilesd.html

### CODE REVIEW

**Comment by SvenDowideit** [comment]
@tianon ? (I havn't tested, but I like the idea)

**Comment by paolomainardi** [comment]
This is just the kernel part, than we need the userspace client: http://linux.die.net/man/8/cachefilesd
As a static binary to be included in the ISO or a tcz package but needs to be created as it does not exist right now.

**Comment by tianon** [comment]
I'm +1 as long as it's compiled as a module and has negligible impact otherwise.

Can't the userspace client run in a container?

**Comment by paolomainardi** [comment]
thanks @tianon, let me check if the userspace client running in a container, very nice idea indeed, i'll back soon.

**Comment by paolomainardi** [comment]
A bit late, but i've finally find some spare time to test it.

Boot2docker built with fscache support (this PR):
```
root@boot2docker:/mnt/sda1/tmp# lsmod | grep cache
cachefiles             32768  0
fscache                36864  1 cachefiles
```

Cachefilesd dockerized Dockerfile:
```
FROM ubuntu:14.04
MAINTAINER Paolo Mainardi "paolo@twinbit.it"
ENV UPDATE_AT 1
RUN apt-get update && apt-get -y install cachefilesd
CMD /sbin/cachefilesd -n -f /etc/cachefilesd.conf -s
```

RUN output:
```
% docker run --privileged -it --rm cachefilesd                                                                                                                                         
About to bind cache
CacheFiles bind failed: errno 95 (Operation not supported)
```

Container dmesg:
```
CacheFiles: Security denies permission to nominate security context: error -95
CacheFiles: Security denies permission to nominate security context: error -95
CacheFiles: Security denies permission to nominate security context: error -95
```

The problem seems related to this function: https://github.com/dz0ny/nims-kernel-ics/blob/master/fs/cachefiles/security.c#L34

I'm still trying to understand what is going on, in the meanwhile if have any suggestions they are more than welcome.

**Comment by paolomainardi** [comment]
Ok, it seems that cachefilesd client requires some selinux kernel configuration, i'm going to put everything together, i'll keep you posted.

**Comment by paolomainardi** [comment]
Updated with required kernel configurations.
I've also tried to run cachefilesd in a container and it works as expected.

File: kernel_config [hunk 1]
```diff
@@ -4187,14 +4196,27 @@ CONFIG_OPTIMIZE_INLINING=y
...
-# CONFIG_SECURITYFS is not set
-# CONFIG_INTEL_TXT is not set
+CONFIG_SECURITY=y
+CONFIG_SECURITYFS=y
+CONFIG_SECURITY_NETWORK=y
+CONFIG_SECURITY_NETWORK_XFRM=y
+CONFIG_SECURITY_PATH=y
+CONFIG_INTEL_TXT=y
+CONFIG_LSM_MMAP_MIN_ADDR=0
+CONFIG_SECURITY_SELINUX=y
```
Comment (tianon) [inline kernel_config]: Won't this have other nasty, unexpected side effects?

Comment (paolomainardi) [inline kernel_config]: Do you mean just for SELINUX or the overall changes ?

Comment (tianon) [inline kernel_config]: I mean just SELinux :cry:

Comment (paolomainardi) [inline kernel_config]: @tianon yes i understand, it was one of my first concern about that, but i tried it extensively without any kind of problem, i think that without the userspace tools it is quite harmless.

**Comment by paolomainardi** [comment]
@tianon any updates ?

**Comment by paolomainardi** [comment]
Sorry for bumping up, any updates @tianon @SvenDowideit ?

**Comment by SvenDowideit** [comment]
adding SELinux brings up red flags which mean we need to do some more research - we discussed it somewhere, and were told there were consequences - but this might be under slightly different details.

**Comment by tianon** [comment]
Indeed, that discussion was in https://github.com/boot2docker/boot2docker/issues/459#issuecomment-52443564.

**Comment by paolomainardi** [comment]
@tianon @SvenDowideit i've tried extensively in the meanwhile without any nasty side effects. Anyway SELINUX is disabled by default:
```
+CONFIG_SECURITY_SELINUX_BOOTPARAM=y
+CONFIG_SECURITY_SELINUX_BOOTPARAM_VALUE=0
+CONFIG_SECURITY_SELINUX_DISABLE=y
```

**Comment by SvenDowideit** [comment]
@paolomainardi can you rebase this PR please?

**Comment by paolomainardi** [comment]
@SvenDowideit yes, i'll do it ASAP.

**Comment by paolomainardi** [comment]
@SvenDowideit rebased and tested, it works as expected.

**Comment by paolomainardi** [comment]
In order to test it, follows this steps.

1) Configure NFS on OSX side.

Edit `/etc/exports` as follows:

```
/Users -mapall=YOURUSER:staff BOOT2DOCKER_IP
```

Replacing `YOURUSER` with the result of `whoami` and BOOT2DOCKER_IP with the result of `boot2docker ip`

Now restart NFS service: `sudo nfsd stop && sudo nfsd start`

2) Create cachefilesd container and load the kernel module:

```
FROM ubuntu:14.04
MAINTAINER Paolo Mainardi "paolo@twinbit.it"
ENV UPDATE_AT 1
RUN apt-get update && apt-get -y install cachefilesd
CMD /sbin/cachefilesd -n -f /etc/cachefilesd.conf -s
```

Compile and run:

```
boot2docker ssh "sudo modprobe cachefiles"
docker build -t cachefilesd .
docker run --privileged -d -v /mnt/sda1/var/cache/fscache:/var/cache/fscache cachefilesd
```

You should see something like this:

```
% docker ps                                                                                                                                                                              
CONTAINER ID        IMAGE                               COMMAND                CREATED              STATUS              PORTS                    NAMES
5cc4416c9f58        cachefilesd:latest                  "/bin/sh -c '/sbin/c   13 minutes ago       Up 13 minutes                                lonely_goldstine
```

3) Unmount vboxfs and mount NFS with `fsc` attribute on:

```
boot2docker ssh 'sudo umount /Users'
boot2docker ssh 'sudo /usr/local/etc/init.d/nfs-client start'
boot2docker ssh 'sudo mount.nfs BOOT2DOCKER_IP:/Users /Users -o fsc,rw,async,noatime,rsize=32768,wsize=32768,proto=tcp'
```

Check if it works correctly, you should see something like this:

```
% boot2docker ssh 'sudo cat /proc/fs/nfsfs/volumes'                                                                                                                             
NV SERVER   PORT DEV     FSID              FSC
v3 c0a83b03  801 0:72    1000003:0         yes
```

**Comment by SvenDowideit** [comment]
I'll do some testing in the next week - doing some conference speaking atm

**Comment by paolomainardi** [comment]
ok @SvenDowideit thanks

**Comment by paolomainardi** [comment]
related to #828 #817 #224

**Comment by SvenDowideit** [comment]
@tianon I can't find a downside to this change.

**Comment by paolomainardi** [comment]
Thanks for testing @SvenDowideit

**Comment by tianon** [comment]
Thanks @SvenDowideit :heart:

**Comment by tianon** [comment]
LGTM

**Comment by SvenDowideit** [comment]
i finally made time to play with the experimental volume plugins - 

`docker run --rm -it --volume-driver=nfs -v 127.0.0.1/data:/no busybox ls -la` works :)

**Comment by thaJeztah** [comment]
@SvenDowideit that's interesting! Looks like your README is incomplete 

> To use the plugin when mounting a volume, run:
> License
> MIT

I tried, but :wink: :
```
root@e0b10a13dcf8:/# License MIT
bash: License: command not found
```

**Comment by paolomainardi** [comment]
Wow, thanks a lot guys!

Now i nave to test new NFS volume driver and we need a simple way to choose vboxfs/NFS share.

**Comment by SvenDowideit** [comment]
@thaJeztah ooops - it was in the Makefile, but i forgot to talk it up :)

I havn't had time since my flight, but hopefully I'll find time during one of the hackathons (or worst case, July, when I'm back in AU.

**Pull request merged**